### ISSUE TITLE
Typeahead ngForm validation proposal

### ISSUE THREAD

**Issue by fernando-sendMail** [angular-ui/bootstrap#3508]
Hi Everyone:

I've been using some of the components in the library, and I've noticed that the typeahead validates right after the user starts typing to look up for an element. I'm unaware of the thought process of why this has been done like this. 

I tried to find any issue open addressing this behavior and I couldn't find any. I would like to make a proposal or start a discussion on how the typeahead should validate or run it's validators.

Currently, when a user starts typing in the `<input>` field, the directive will start looking up for any match for whatever the user is typing, also at the same time, validators are fired at that point, this validators affect, if present, the state of any ngForm reference. Assume we have this:

``` 
<form name="user">
  ...
  <input type="text" name="name" typeahead="..." typeahead-editable="false">
  ...
  <div ng-show="!user.name.$valid">
    Invalid name for user!
  </div>
</form>
``` 
As soon as the user starts typing a value the message will start showing even though that doesn't mean the user has not selected an incorrect value yet. In my opinion validation for this should happen after the user has selected a value by any means (clicking, hitting enter, etc) or if the `<input>` element looses the focus, this would create a more meaningful user experience when using this component, rather than setting the value of $valid to false right away.

Right now I created a directive to detect when the blur event is fired from the typeahead element so the error message actually shows when the user has lost focus on that element, implying that either the user left the element blank because he thought it was not a required field or because he typed a valid value.

I hope we can get to improve the existing code base and usability of this element, if you have any comments please let me know.

Thanks!

**Comment by rvanbaalen** [comment]
If you're using AngularJS 1.3, I think you can work around this behavior by setting your `ng-model-options`

```js
<input type="text" ng-model="model.name" ng-model-options="{'updateOn': 'blur'}">
```

If you can setup a plunkr demonstrating your issue, I'll test of this fixes it.

**Comment by fernando-sendMail** [comment]
Thanks for the info @rvanbaalen I was aware of that new feature and I tried it before, the problem is that if you are using typeahead with promises (`$http`) it doesn't show the results until you blur that element.

Here's a plunkr with that issue:

http://plnkr.co/edit/TEIy8qofsVO87q40IZSj?p=preview

As you can see in the first typeahead the results wont popup until the typeahead input looses focus. On the second one the typeahead shows the results but it's marked as invalid when the user starts typing.

**Comment by rvanbaalen** [comment]
There are no typeaheads in your plunker @fernando-sendMail. Perhaps you forgot to save.

Also, when checking for validation, always check for both `dirty` (or !pristine) **and** error. In the plunker you showed you only set the `.ng-invalid` style.

**Comment by fernando-sendMail** [comment]
Sorry about that @rvanbaalen, now it's working and yes, I also checked for that in this version.

http://plnkr.co/edit/TEIy8qofsVO87q40IZSj?p=preview

**Comment by fernando-sendMail** [comment]
@rvanbaalen I tried the approach of the checking the dirty but the issue is that angular sets the dirty just after you start typing and also the $error object, this makes perfect sense when styling just the input element, but not for other elements such as inline error messages in the form.

@karianna Is there any specific reason on why this was closed? (In the previous comment I meant that the plunkr link was fixed).

**Comment by rvanbaalen** [comment]
I think @karianna was a bit trigger happy :wink:

I'll close this when it's done

**Comment by rvanbaalen** [comment]
@fernando-sendMail the behavior you describe is not something we came up with its default Angular model behavior. Like I said before you can address this by manipulating the ng-model-options.

**Comment by rvanbaalen** [comment]
I managed to reproduce your problem in a seperate plunkr indeed.

http://plnkr.co/edit/BvHemSy1iaADwGkPU1Ly

This problem only exists when using the typeahead directive in combination with `typeahead-editable=false`. It's related to the way Angular 1.3 handles model values.

**Comment by chrisirhc** [comment]
I think the bug here is that when `ng-model-options="{'updateOn': 'blur'}"` is used, the popup only appears on blur. This looks definitely looks like it needs to be fixed.

**Comment by RobJacobs** [comment]
I think this is related to PR #3187

**Comment by rvanbaalen** [comment]
Good find @RobJacobs. This one is related as well:

https://github.com/angular-ui/bootstrap/issues/3219

**Comment by fernando-sendMail** [comment]
Thanks for your attention to this issue, do you know if somebody is working on this issue? (I couldn't find anyone that is working on this)

If not I might be able to work on a solution for this problem and create a PR.

**Comment by rvanbaalen** [comment]
As you can see @fernando-sendMail, @RobJacobs already pointed to a PR. You can go ahead and verify if that PR fixes your issue as well.

**Comment by rvanbaalen** [comment]
@fernando-sendMail My bad, @RobJacobs called it a PR but it's a related issue, not a PR.

**Comment by fernando-sendMail** [comment]
Yes haha, don't worry, I can see that someone is trying to work on that but he/she just referenced the issue and didn't create any PRs or test for the solution. I'll try to look at the problem and see what I can do.

**Comment by jehoshua02** [comment]
I don't know if this is related, but we are using Angular 1.4.8, and angular-ui-bootstrap 0.14.3. The issue we were having is the typeahead was allowing values outside the list fed to typeahead. After banging my head for an hour, and looking at the typeahead docs again, I noticed the code examples did NOT prefix the typeahead attributes with `uib-`. Removing these prefixes solved our problem.

http://plnkr.co/edit/CK9rUM?p=preview
